// @c8y/ngx-components/esm5/core/provider-configuration/provider-configuration-route.factory.js
import * as tslib_1 from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { flatten } from 'lodash-es';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
import { ProviderConfigurationComponent } from './provider-configuration.component';
var ProviderConfigurationRouteFactory = /** @class */ (function () {
    function ProviderConfigurationRouteFactory(config) {
        this.config = flatten(config);
    }
    ProviderConfigurationRouteFactory.prototype.get = function () {
        return this.config.map(function (c) { return ({
            path: c.navigation.path,
            component: ProviderConfigurationComponent,
            canActivate: c.navigation.canActivate,
            data: {
                layout: c.layout,
                endpoint: c.endpoint
            }
        }); });
    };
    ProviderConfigurationRouteFactory.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_DYNAMIC_PROVIDER_CONFIG,] }] }
    ]; };
    ProviderConfigurationRouteFactory = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(0, Optional()), tslib_1.__param(0, Inject(HOOK_DYNAMIC_PROVIDER_CONFIG))
    ], ProviderConfigurationRouteFactory);
    return ProviderConfigurationRouteFactory;
}());
export { ProviderConfigurationRouteFactory };
