// @c8y/ngx-components/esm5/core/provider-configuration/provider-configuration-node.factory.js
import * as tslib_1 from "tslib";
import { Inject, Injectable, Injector, Optional } from '@angular/core';
import { flatten } from 'lodash-es';
import { forkJoin, from, isObservable, of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { NavigatorNode } from '../navigator';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
var ProviderConfigurationNodeFactory = /** @class */ (function () {
    function ProviderConfigurationNodeFactory(config, injector) {
        this.injector = injector;
        this.config = flatten(config);
    }
    ProviderConfigurationNodeFactory.prototype.get = function () {
        var _this = this;
        if (!this.config || !this.config.length) {
            return;
        }
        if (!this.nodes) {
            var canActivate = this.config
                .map(function (c) {
                return c.navigation.canActivate && c.navigation.canActivate.length
                    ? c.navigation.canActivate.map(function (ca) { return _this.injector.get(ca); })
                    : undefined;
            })
                .map(this.checkCanActivate.bind(this));
            return forkJoin(canActivate).pipe(map(function (canActivateResult) {
                return _this.config
                    .map(function (c, index) {
                    return canActivateResult[index] ? new NavigatorNode(c.navigation) : undefined;
                })
                    .filter(function (el) { return !!el; });
            }), tap(function (nodes) { return (_this.nodes = nodes); }));
        }
        return this.nodes;
    };
    ProviderConfigurationNodeFactory.prototype.checkCanActivate = function (ca) {
        if (!!ca && ca.length) {
            var canActivateResult = ca
                .map(function (canActivate) { return canActivate.canActivate(undefined, undefined); })
                .map(this.wrapIntoObservable.bind(this));
            return forkJoin(canActivateResult).pipe(map(function (caResult) { return caResult.reduce(function (acc, curr) { return acc && curr; }); }));
        }
        return of(true);
    };
    ProviderConfigurationNodeFactory.prototype.isPromise = function (obj) {
        return !!obj && typeof obj.then === 'function';
    };
    ProviderConfigurationNodeFactory.prototype.wrapIntoObservable = function (value) {
        if (isObservable(value)) {
            return value;
        }
        if (this.isPromise(value)) {
            return from(value);
        }
        return of(value);
    };
    ProviderConfigurationNodeFactory.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_DYNAMIC_PROVIDER_CONFIG,] }] },
        { type: Injector }
    ]; };
    ProviderConfigurationNodeFactory = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(0, Optional()),
        tslib_1.__param(0, Inject(HOOK_DYNAMIC_PROVIDER_CONFIG))
    ], ProviderConfigurationNodeFactory);
    return ProviderConfigurationNodeFactory;
}());
export { ProviderConfigurationNodeFactory };
