// @c8y/ngx-components/esm5/operations/bulk/creation/bulk-operations-stepper-container.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, ViewContainerRef, ViewChild, ComponentFactoryResolver, ChangeDetectorRef } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
var BulkOperationsStepperContainerComponent = /** @class */ (function () {
    function BulkOperationsStepperContainerComponent(activatedRoute, bulkOpsService, resolver, changeDetector) {
        this.activatedRoute = activatedRoute;
        this.bulkOpsService = bulkOpsService;
        this.resolver = resolver;
        this.changeDetector = changeDetector;
    }
    BulkOperationsStepperContainerComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.activatedRoute.params.subscribe(function (params) {
            if (params && params.type) {
                _this.bulkOperationType = params.type;
                _this.component = _this.bulkOpsService
                    .getBulkTypes()
                    .find(function (element) { return element.path.includes(_this.bulkOperationType); }).component;
                _this.changeDetector.detectChanges();
                _this.createComponent(_this.component);
            }
        });
    };
    BulkOperationsStepperContainerComponent.prototype.createComponent = function (component) {
        if (component) {
            var factory = this.resolver.resolveComponentFactory(component);
            this.container.createComponent(factory);
        }
    };
    BulkOperationsStepperContainerComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: BulkOperationsService },
        { type: ComponentFactoryResolver },
        { type: ChangeDetectorRef }
    ]; };
    tslib_1.__decorate([
        ViewChild('container', { static: false, read: ViewContainerRef })
    ], BulkOperationsStepperContainerComponent.prototype, "container", void 0);
    BulkOperationsStepperContainerComponent = tslib_1.__decorate([
        Component({
            selector: 'bulk-operations-stepper-container',
            template: "<c8y-title>{{ 'New bulk operation' | translate }}</c8y-title>\n<div class=\"card card--grid--fullpage\">\n  <ng-container [ngSwitch]=\"bulkOperationType\">\n    <!-- Cases: software, firmware and configuration are default c8y-types -->\n    <div *ngSwitchCase=\"'software'\" class=\"d-contents\">\n      <c8y-stepper-bulk-type-software></c8y-stepper-bulk-type-software>\n    </div>\n    <div *ngSwitchCase=\"'firmware'\" class=\"d-contents\">\n      <c8y-stepper-firmware></c8y-stepper-firmware>\n    </div>\n    <div *ngSwitchCase=\"'configuration'\" class=\"d-contents\">\n      <c8y-stepper-bulk-type-configuration></c8y-stepper-bulk-type-configuration>\n    </div>\n    <div *ngSwitchCase=\"'device-profile'\" class=\"d-contents\">\n      <c8y-stepper-bulk-type-device-profile></c8y-stepper-bulk-type-device-profile>\n    </div>\n    <!-- This case is for 3rd-party-dev defined types -->\n    <div *ngSwitchDefault class=\"d-contents\">\n      <template #container></template>\n    </div>\n  </ng-container>\n</div>\n"
        })
    ], BulkOperationsStepperContainerComponent);
    return BulkOperationsStepperContainerComponent;
}());
export { BulkOperationsStepperContainerComponent };
