// @c8y/ngx-components/esm2015/operations/bulk/create-from-single/bulk-operation-from-single.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { get } from 'lodash-es';
import { BaseStepperComponent } from '../base-stepper.component';
import { BulkOperationsService, OperationDetails } from '@c8y/ngx-components/operations/bulk-operations-service';
let BulkOperationFromSingleComponent = class BulkOperationFromSingleComponent extends BaseStepperComponent {
    constructor(activatedRoute, bulkOperationService) {
        super();
        this.activatedRoute = activatedRoute;
        this.bulkOperationService = bulkOperationService;
    }
    retrieveOperationPrototype() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const prototype = (yield this.bulkOperationService.getOperation(this.activatedRoute.snapshot.params.operationId)).data;
            return {
                description: get(prototype, 'description'),
                prototype
            };
        });
    }
};
BulkOperationFromSingleComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: BulkOperationsService }
];
BulkOperationFromSingleComponent = tslib_1.__decorate([
    Component({
        selector: 'bulk-operation-from-single',
        template: "<c8y-bulk-operation-stepper> </c8y-bulk-operation-stepper>\n"
    })
], BulkOperationFromSingleComponent);
export { BulkOperationFromSingleComponent };
