// @c8y/ngx-components/esm5/core/dynamic-forms/dynamic-forms-ui-theme.module.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule, FORMLY_CONFIG } from '@ngx-formly/core';
import { FormlySelectModule } from '@ngx-formly/core/select';
import { TranslateModule } from '@ngx-translate/core';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { FormsModule } from '../forms/index';
import { ArrayTypeComponent } from './array/array.type.component';
import { FieldCheckbox } from './checkbox/checkbox.type.component';
import { WrapperFormField } from './form-field/form-field.wrapper.component';
import { FieldInput } from './input/input.type.component';
import { ObjectTypeComponent } from './object/object.type.component';
import { FieldRadio } from './radio/radio.type.component';
export var FIELD_TYPES = [
    FieldCheckbox,
    ArrayTypeComponent,
    WrapperFormField,
    FieldInput,
    FieldRadio,
    ObjectTypeComponent
];
export function fieldTypeFactory() {
    // tslint:disable-next-line:no-string-literal
    return FIELD_TYPES.map(function (type) { return type['CONFIG']; }).reduce(function (formlyConfig, typeConfig) {
        return (tslib_1.__assign({}, formlyConfig, { types: tslib_1.__spread(formlyConfig.types, (typeConfig.types || [])), wrappers: tslib_1.__spread(formlyConfig.wrappers, (typeConfig.wrappers || [])) }));
    }, { types: [], wrappers: [] });
}
/**
 * This module declares field type components for dynamic forms that match C8Y styleguide.
 *
 * @dynamic
 */
var DynamicFormsUIThemeModule = /** @class */ (function () {
    function DynamicFormsUIThemeModule() {
    }
    DynamicFormsUIThemeModule = tslib_1.__decorate([
        NgModule({
            declarations: tslib_1.__spread(FIELD_TYPES),
            entryComponents: tslib_1.__spread(FIELD_TYPES),
            imports: [
                CommonModule,
                PopoverModule,
                ReactiveFormsModule,
                TranslateModule,
                FormsModule,
                FormlyModule,
                FormlySelectModule
            ],
            providers: [
                {
                    provide: FORMLY_CONFIG,
                    useFactory: fieldTypeFactory,
                    multi: true
                }
            ]
        })
    ], DynamicFormsUIThemeModule);
    return DynamicFormsUIThemeModule;
}());
export { DynamicFormsUIThemeModule };
