// @c8y/ngx-components/esm5/core/bootstrap/cookie-banner/cookie-banner.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { CookieBannerPreferencesModalComponent } from './cookie-banner-preferences-modal/cookie-banner-preferences-modal.component';
import { CookieBannerService } from './cookie-banner.service';
var CookieBannerComponent = /** @class */ (function () {
    function CookieBannerComponent(bsModalService, cookieBannerService) {
        this.bsModalService = bsModalService;
        this.cookieBannerService = cookieBannerService;
        this.shouldDisable = false;
    }
    CookieBannerComponent.prototype.ngOnInit = function () {
        this.settings = this.cookieBannerService.getCookieBannerSettings();
        this.cookiePreferences = this.cookieBannerService.getCookiePreferencesConfig();
        this.shouldShowCookieBanner();
    };
    CookieBannerComponent.prototype.onConfigurePreferences = function () {
        var _this = this;
        this.disableButtons(true);
        var initialState = {
            configCookiePreferences: this.cookiePreferences
        };
        var modalRef = this.bsModalService.show(CookieBannerPreferencesModalComponent, {
            initialState: initialState
        });
        modalRef.content.closeSubject.subscribe(function (userCookiePreferences) {
            _this.disableButtons(false);
            if (userCookiePreferences) {
                _this.cookieBannerService.setCookies(userCookiePreferences);
                _this.setCookieBannerVisibility(false);
            }
        });
    };
    CookieBannerComponent.prototype.acceptCookies = function () {
        this.cookieBannerService.setCookies(this.cookiePreferences || { required: true });
        this.setCookieBannerVisibility(false);
    };
    CookieBannerComponent.prototype.shouldShowCookieBanner = function () {
        this.setCookieBannerVisibility(this.cookieBannerService.shouldShowCookieBanner());
    };
    CookieBannerComponent.prototype.disableButtons = function (shouldDisable) {
        this.shouldDisable = shouldDisable;
    };
    CookieBannerComponent.prototype.setCookieBannerVisibility = function (shouldShow) {
        this.showBanner = shouldShow;
        this.cookieBannerService.isCookieBannerShowed$.next(shouldShow);
    };
    CookieBannerComponent.ctorParameters = function () { return [
        { type: BsModalService },
        { type: CookieBannerService }
    ]; };
    CookieBannerComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-cookie-banner',
            template: "<div class=\"c8y-cookie-banner\">\n  <div class=\"container\">\n    <div class=\"row p-t-16 p-b-16\" *ngIf=\"showBanner\">\n      <div class=\"col-md-2\" *ngIf=\"settings.cookieBannerTitle\">\n        <h3>{{ settings.cookieBannerTitle | translate }}</h3>\n        <div class=\"p-b-16 visible-sm visible-xs\"></div>\n      </div>\n      <div class=\"col-md-7\">\n        <div class=\"cookie-banner-content text-pre\">\n          {{ settings.cookieBannerText | translate }}\n        </div>\n      </div>\n      <div class=\"col-md-3\">\n        <div class=\"row\">\n          <div class=\"col-sm-6 col-md-12\">\n            <div class=\"p-b-16 visible-sm visible-xs\"></div>\n            <button\n              class=\"btn btn-primary btn-block\"\n              (click)=\"acceptCookies()\"\n              [disabled]=\"shouldDisable\"\n            >\n              {{ 'Agree and proceed' | translate }}\n            </button>\n          </div>\n          <div class=\"col-sm-6 col-md-12\" *ngIf=\"!!cookiePreferences\">\n            <div class=\"p-b-16\"></div>\n            <button class=\"btn btn-default btn-block\" (click)=\"onConfigurePreferences()\" [disabled]=\"shouldDisable\">\n              {{ 'Configure preferences' | translate }}\n            </button>\n          </div>\n          <div class=\"col-sm-6 col-md-12\">\n            <div class=\"p-b-16\"></div>\n            <a\n              class=\"btn btn-default btn-block\"\n              href=\"{{ settings.policyUrl }}\"\n              target=\"_blank\"\n              *ngIf=\"settings.policyUrl\"\n            >\n              {{ 'Privacy policy' | translate }}\n            </a>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n"
        })
    ], CookieBannerComponent);
    return CookieBannerComponent;
}());
export { CookieBannerComponent };
