// @c8y/ngx-components/esm5/core/bootstrap/cookie-banner/cookie-banner.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { OptionsService } from '../../common/options.service';
import { gettext } from '../../i18n/gettext';
import * as i0 from "@angular/core";
import * as i1 from "../../common/options.service";
var CookieBannerService = /** @class */ (function () {
    function CookieBannerService(options) {
        this.options = options;
        this.STORAGE_KEY = 'acceptCookieNotice';
        this.COOKIE_BANNER = 'cookieBanner';
        this.isCookieBannerShowed$ = new Subject();
        this.cookieDescriptions = {
            required: gettext("These cookies are required to enable core site functionality. They perform a task or operation without which a site's functionality would not be possible."),
            functional: gettext('These cookies are used to track site usage and to process my personal data to measure and improve usability and performance. We will neither forward this data to any third party nor use this data for direct marketing purposes. I recognize that I can opt in and out of these cookies at any time.'),
            marketing: gettext('These cookies are used to target advertising to a user.')
        };
    }
    CookieBannerService.prototype.getCookiePreferencesConfig = function () {
        var settings = this.getCookieBannerSettings();
        return settings && settings.configuration;
    };
    CookieBannerService.prototype.getCookieBannerSettings = function () {
        return this.options.get(this.COOKIE_BANNER) || {};
    };
    CookieBannerService.prototype.setCookies = function (cookiePreferences) {
        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(cookiePreferences));
    };
    CookieBannerService.prototype.shouldShowCookieBanner = function () {
        var settings = this.getCookieBannerSettings();
        return !!settings.cookieBannerText && !this.getUserCookiePreferences();
    };
    CookieBannerService.prototype.getUserCookiePreferences = function () {
        return JSON.parse(localStorage.getItem(this.STORAGE_KEY));
    };
    CookieBannerService.prototype.isConfigCookiePreferencesDefined = function () {
        return !!this.getCookiePreferencesConfig();
    };
    CookieBannerService.prototype.isFunctionalCookieDisabled = function () {
        var functional = (this.getUserCookiePreferences() || {}).functional;
        /**
         * We want this check, as we should stop loading Gainsight tag only when functional cookie is explicitly set to false.
         * In the USA there is no need for the functional cookies when tracking users.
         * This is why we want to allow undefined to pass the check.
         * @ignore
         */
        return functional === false;
    };
    CookieBannerService.prototype.getCookieDescription = function (cookieType) {
        return this.cookieDescriptions[cookieType];
    };
    CookieBannerService.prototype.transformCookiePreferencesToList = function (cookiePreferences) {
        return Object.entries(cookiePreferences).map(function (_a) {
            var _b = tslib_1.__read(_a, 2), name = _b[0], value = _b[1];
            var isReadonly = name === 'required';
            return { name: name, value: value, isReadonly: isReadonly };
        });
    };
    CookieBannerService.prototype.transformCookiePreferencesListToCookiePreferences = function (cookiePreferencesList) {
        return cookiePreferencesList.reduce(function (cookiePref, cookie) {
            var key = cookie.name;
            cookiePref[key] = cookie.value;
            return cookiePref;
        }, {});
    };
    CookieBannerService.ctorParameters = function () { return [
        { type: OptionsService }
    ]; };
    CookieBannerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CookieBannerService_Factory() { return new CookieBannerService(i0.ɵɵinject(i1.OptionsService)); }, token: CookieBannerService, providedIn: "root" });
    CookieBannerService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], CookieBannerService);
    return CookieBannerService;
}());
export { CookieBannerService };
