// @c8y/ngx-components/esm2015/core/navigator/navigator-outlet.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { HeaderService } from '../header/index';
import { BehaviorSubject } from 'rxjs';
let NavigatorOutletComponent = class NavigatorOutletComponent {
    constructor(headerService) {
        this.headerService = headerService;
        this.open = true;
        this.isShown$ = new BehaviorSubject(false);
    }
    ngOnChanges() {
        this.isShown$.next(this.nodes && this.nodes.length > 0);
    }
    /**
     * Closes the navigator on click if the window is small.
     */
    collapseNavigator(from) {
        if (from === 'link' && !this.headerService.largeWidth) {
            this.headerService.closeNavigator();
        }
    }
    trackByPathOrLabel(index, node) {
        return node.path ? node.path : node.label;
    }
};
NavigatorOutletComponent.ctorParameters = () => [
    { type: HeaderService }
];
tslib_1.__decorate([
    Input()
], NavigatorOutletComponent.prototype, "nodes", void 0);
tslib_1.__decorate([
    Input()
], NavigatorOutletComponent.prototype, "open", void 0);
NavigatorOutletComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-navigator-outlet, c8y-ui-navigator',
        template: "<nav class=\"navigator\" *ngIf=\"isShown$ | async\" [ngClass]=\"{ open: open }\">\n  <ng-content select=\".navigator-slot-top\"></ng-content>\n  <ul class=\"navigatorContent\">\n    <c8y-navigator-node-display (nodeClick)=\"collapseNavigator($event)\" [node]=\"node\" *ngFor=\"let node of nodes trackBy trackByPathOrLabel\" [isRoot]=\"true\"></c8y-navigator-node-display>\n  </ul>\n  <ng-content select=\".navigator-slot-bottom\"></ng-content>\n</nav>\n"
    })
], NavigatorOutletComponent);
export { NavigatorOutletComponent };
