// @c8y/ngx-components/esm5/upgrade/ng1/downgraded.services.js
import { downgradeInjectable } from '@angular/upgrade/static';
import { AlertService, AppStateService, DocsService, HeaderService, PasswordService, UserMenuService, CachedLocaleDictionaryService } from '@c8y/ngx-components';
import { ApiService } from '@c8y/ngx-components/api';
import { BulkOperationDetailsService } from '@c8y/ngx-components/operations/bulk-operations-list';
import { TranslateService } from '@ngx-translate/core';
import { BridgeService } from '../bridge.service';
import { ServerMessagesService } from './serverMessages.service';
import { AuthBridgeService } from '../auth-bridge.service';
export var bridgeService = downgradeInjectable(BridgeService);
export var authBridgeService = downgradeInjectable(AuthBridgeService);
export var appStateService = downgradeInjectable(AppStateService);
export var headerService = downgradeInjectable(HeaderService);
export var alertsService = downgradeInjectable(AlertService);
export var userMenuService = downgradeInjectable(UserMenuService);
export var apiService = downgradeInjectable(ApiService);
export var docsService = downgradeInjectable(DocsService);
export var passwordService = downgradeInjectable(PasswordService);
export var translateService = downgradeInjectable(TranslateService);
export var cachedLocaleDictionaryService = downgradeInjectable(CachedLocaleDictionaryService);
export var serverMessagesService = downgradeInjectable(ServerMessagesService);
export var bulkOperationDetailsService = downgradeInjectable(BulkOperationDetailsService);
