// @c8y/ngx-components/esm5/core/docs/guide-docs.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, ContentChildren, QueryList, ContentChild, AfterContentInit, ElementRef, Renderer2, OnDestroy } from '@angular/core';
import { GuideHrefDirective } from './guide-href.directive';
import { DocsService } from './docs.service';
import { ReplaySubject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { ForceHtmlTranslateDirective } from '../i18n/force-html-translate.directive';
var GuideDocsComponent = /** @class */ (function () {
    function GuideDocsComponent(docs, el, renderer) {
        this.docs = docs;
        this.el = el;
        this.renderer = renderer;
        this.destroyed$ = new ReplaySubject(1);
    }
    GuideDocsComponent.prototype.ngOnInit = function () {
        this.baseUrl = this.docs.getBaseUrl();
    };
    /**
     * After content is initialized we subscribe for translation events, then we are able to modify
     * anchor just as C8yGuideHrefDirective does.
     */
    GuideDocsComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (this.translateDirective) {
            this.translateDirective.htmlUpdateEvent
                .pipe(takeUntil(this.destroyed$))
                .subscribe(function () {
                _this.replaceAnchorAttributes();
            });
        }
        else {
            this.replaceAnchorAttributes();
        }
    };
    /**
     * Find and modify anchor elements
     */
    GuideDocsComponent.prototype.replaceAnchorAttributes = function () {
        var _this = this;
        var anchors = this.findGuideAnchors(this.el.nativeElement);
        anchors.forEach(function (a) {
            var guideHref = _this.baseUrl + "/" + a.getAttribute('c8y-guide-href');
            _this.renderer.setAttribute(a, 'href', guideHref);
            if (!a.hasAttribute('target')) {
                _this.renderer.setAttribute(a, 'target', '_blank');
            }
        });
    };
    /**
     * Recursive function, returns array of Anchor elements inside HTMLElement.
     */
    GuideDocsComponent.prototype.findGuideAnchors = function (el) {
        var _this = this;
        var isAnchor = function () { return el.nodeName === 'A'; };
        var hasGuideHrefAttr = function () { return !!el.attributes.getNamedItem('c8y-guide-href'); };
        var hasChildren = function () { return el.childNodes && el.childNodes.length > 0; };
        var result = [];
        if (isAnchor() && hasGuideHrefAttr()) {
            result.push(el);
        }
        else if (hasChildren()) {
            el.childNodes.forEach(function (node) {
                var searchResult = _this.findGuideAnchors(node);
                if (searchResult) {
                    result = tslib_1.__spread(result, searchResult);
                }
            });
        }
        return result;
    };
    GuideDocsComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.links.forEach(function (guide) {
            guide.baseHref = _this.baseUrl;
        });
    };
    GuideDocsComponent.prototype.ngOnDestroy = function () {
        this.destroyed$.next(true);
        this.destroyed$.complete();
    };
    GuideDocsComponent.ctorParameters = function () { return [
        { type: DocsService },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        ContentChildren(GuideHrefDirective, { descendants: true })
    ], GuideDocsComponent.prototype, "links", void 0);
    tslib_1.__decorate([
        ContentChild(ForceHtmlTranslateDirective, { static: false })
    ], GuideDocsComponent.prototype, "translateDirective", void 0);
    GuideDocsComponent = tslib_1.__decorate([
        Component({
            selector: '[c8y-guide-docs]',
            template: "<ng-container *ngIf=\"!!baseUrl\">\n  <ng-content></ng-content>\n</ng-container>\n"
        })
    ], GuideDocsComponent);
    return GuideDocsComponent;
}());
export { GuideDocsComponent };
