// @c8y/ngx-components/esm2015/core/bootstrap/cookie-banner/cookie-banner.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { OptionsService } from '../../common/options.service';
import { gettext } from '../../i18n/gettext';
import * as i0 from "@angular/core";
import * as i1 from "../../common/options.service";
let CookieBannerService = class CookieBannerService {
    constructor(options) {
        this.options = options;
        this.STORAGE_KEY = 'acceptCookieNotice';
        this.COOKIE_BANNER = 'cookieBanner';
        this.isCookieBannerShowed$ = new Subject();
        this.cookieDescriptions = {
            required: gettext("These cookies are required to enable core site functionality. They perform a task or operation without which a site's functionality would not be possible."),
            functional: gettext('These cookies are used to track site usage and to process my personal data to measure and improve usability and performance. We will neither forward this data to any third party nor use this data for direct marketing purposes. I recognize that I can opt in and out of these cookies at any time.'),
            marketing: gettext('These cookies are used to target advertising to a user.')
        };
    }
    getCookiePreferencesConfig() {
        const settings = this.getCookieBannerSettings();
        return settings && settings.configuration;
    }
    getCookieBannerSettings() {
        return this.options.get(this.COOKIE_BANNER) || {};
    }
    setCookies(cookiePreferences) {
        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(cookiePreferences));
    }
    shouldShowCookieBanner() {
        const settings = this.getCookieBannerSettings();
        return !!settings.cookieBannerText && !this.getUserCookiePreferences();
    }
    getUserCookiePreferences() {
        return JSON.parse(localStorage.getItem(this.STORAGE_KEY));
    }
    isConfigCookiePreferencesDefined() {
        return !!this.getCookiePreferencesConfig();
    }
    isFunctionalCookieDisabled() {
        const { functional } = this.getUserCookiePreferences() || {};
        /**
         * We want this check, as we should stop loading Gainsight tag only when functional cookie is explicitly set to false.
         * In the USA there is no need for the functional cookies when tracking users.
         * This is why we want to allow undefined to pass the check.
         * @ignore
         */
        return functional === false;
    }
    getCookieDescription(cookieType) {
        return this.cookieDescriptions[cookieType];
    }
    transformCookiePreferencesToList(cookiePreferences) {
        return Object.entries(cookiePreferences).map(([name, value]) => {
            const isReadonly = name === 'required';
            return { name, value, isReadonly };
        });
    }
    transformCookiePreferencesListToCookiePreferences(cookiePreferencesList) {
        return cookiePreferencesList.reduce((cookiePref, cookie) => {
            const key = cookie.name;
            cookiePref[key] = cookie.value;
            return cookiePref;
        }, {});
    }
};
CookieBannerService.ctorParameters = () => [
    { type: OptionsService }
];
CookieBannerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CookieBannerService_Factory() { return new CookieBannerService(i0.ɵɵinject(i1.OptionsService)); }, token: CookieBannerService, providedIn: "root" });
CookieBannerService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    })
], CookieBannerService);
export { CookieBannerService };
