// @c8y/ngx-components/esm5/operations/bulk/create-from-single/bulk-operation-from-single.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { get } from 'lodash-es';
import { BaseStepperComponent } from '../base-stepper.component';
import { BulkOperationsService, OperationDetails } from '@c8y/ngx-components/operations/bulk-operations-service';
var BulkOperationFromSingleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BulkOperationFromSingleComponent, _super);
    function BulkOperationFromSingleComponent(activatedRoute, bulkOperationService) {
        var _this = _super.call(this) || this;
        _this.activatedRoute = activatedRoute;
        _this.bulkOperationService = bulkOperationService;
        return _this;
    }
    BulkOperationFromSingleComponent.prototype.retrieveOperationPrototype = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var prototype;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.bulkOperationService.getOperation(this.activatedRoute.snapshot.params.operationId)];
                    case 1:
                        prototype = (_a.sent()).data;
                        return [2 /*return*/, {
                                description: get(prototype, 'description'),
                                prototype: prototype
                            }];
                }
            });
        });
    };
    BulkOperationFromSingleComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: BulkOperationsService }
    ]; };
    BulkOperationFromSingleComponent = tslib_1.__decorate([
        Component({
            selector: 'bulk-operation-from-single',
            template: "<c8y-bulk-operation-stepper> </c8y-bulk-operation-stepper>\n"
        })
    ], BulkOperationFromSingleComponent);
    return BulkOperationFromSingleComponent;
}(BaseStepperComponent));
export { BulkOperationFromSingleComponent };
