// @c8y/ngx-components/esm5/core/product-experience/product-experience.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { combineLatest } from 'rxjs';
import { filter } from 'rxjs/operators';
import { CookieBannerService } from '../bootstrap/cookie-banner/cookie-banner.service';
import { AppStateService } from '../common/ui-state.service';
import { GainsightService } from './gainsight.service';
import { ProductExperienceDirective } from './product-experience.directive';
/**
 * This module enables an tenant to activate the product experience
 * software [Gainsight](https://www.gainsight.com/product-experience/) to help
 * and track user actions. Gainsight is only activated, if the tenant custom
 * property `gainsightEnabled` is set to true.
 */
var ProductExperienceModule = /** @class */ (function () {
    function ProductExperienceModule(appState, gainsightService, cookieBannerService) {
        var _this = this;
        this.appState = appState;
        this.gainsightService = gainsightService;
        this.cookieBannerService = cookieBannerService;
        combineLatest([
            this.appState.currentTenant.pipe(filter(Boolean)),
            this.cookieBannerService.isCookieBannerShowed$
        ]).subscribe(function (_a) {
            var _b = tslib_1.__read(_a, 2), currentTenant = _b[0], isCookieBannerShowed = _b[1];
            if (isCookieBannerShowed) {
                return;
            }
            var customProperties = currentTenant.customProperties, name = currentTenant.name;
            if (_this.gainsightService.shouldLoadGainsightTag(customProperties)) {
                _this.gainsightService.loadTag(name);
            }
        });
    }
    ProductExperienceModule.ctorParameters = function () { return [
        { type: AppStateService },
        { type: GainsightService },
        { type: CookieBannerService }
    ]; };
    ProductExperienceModule = tslib_1.__decorate([
        NgModule({
            declarations: [ProductExperienceDirective],
            exports: [ProductExperienceDirective],
            providers: [GainsightService]
        })
    ], ProductExperienceModule);
    return ProductExperienceModule;
}());
export { ProductExperienceModule };
