// @c8y/ngx-components/esm5/device-profile/device-tab-profile/device-tab-profile.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { IManagedObject, IOperation, IResultList, Realtime } from '@c8y/client';
import { ActivatedRoute } from '@angular/router';
import { pipe } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { DeviceProfileService } from '../device-profile.service';
import { AlertService, gettext, ManagedObjectRealtimeService } from '@c8y/ngx-components';
var DeviceTabProfileComponent = /** @class */ (function () {
    function DeviceTabProfileComponent(deviceRealtime, deviceProfileService, route, realtime, alertService) {
        this.deviceRealtime = deviceRealtime;
        this.deviceProfileService = deviceProfileService;
        this.route = route;
        this.realtime = realtime;
        this.alertService = alertService;
        this.firmwareItems = [];
        this.softwareItems = [];
        this.configurationItems = [];
        this.pattern = '';
        this.reloading = false;
    }
    DeviceTabProfileComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.device = this.route.snapshot.parent.data.contextData;
                this.getDeviceProfilesAndUpdateProfileItems();
                this.subscribeToManagedObjects();
                return [2 /*return*/];
            });
        });
    };
    DeviceTabProfileComponent.prototype.getDeviceProfilesAndUpdateProfileItems = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, profileId_1, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this.reloading = true;
                        _a = this;
                        return [4 /*yield*/, this.deviceProfileService.getDeviceProfilesByDeviceType(this.device.type)];
                    case 1:
                        _a.deviceProfiles = _c.sent();
                        if (this.device.c8y_Profile) {
                            profileId_1 = this.device.c8y_Profile.profileId;
                            this.selectedProfile = this.deviceProfiles.data.find(function (mo) { return mo.id === profileId_1; });
                        }
                        this.updateProfileItems(this.device, this.selectedProfile);
                        _b = this;
                        return [4 /*yield*/, this.deviceProfileService.getProfileOperation(this.device.id)];
                    case 2:
                        _b.operation = _c.sent();
                        this.subscribeToOperations();
                        this.reloading = false;
                        return [2 /*return*/];
                }
            });
        });
    };
    DeviceTabProfileComponent.prototype.selectProfile = function (mo) {
        this.selectedProfile = mo;
        this.updateProfileItems(this.device, this.selectedProfile);
    };
    DeviceTabProfileComponent.prototype.createOperation = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.deviceProfileService.createProfileOperation(this.device, this.selectedProfile)];
                    case 1:
                        _a.operation = _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    DeviceTabProfileComponent.prototype.setPipe = function (filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map(function (data) {
            return data.filter(function (mo) { return mo.name && mo.name.toLowerCase().indexOf(filterStr.toLowerCase()) > -1; });
        }));
    };
    DeviceTabProfileComponent.prototype.ngOnDestroy = function () {
        this.operationsSubscription.unsubscribe();
        this.moOnUpdateSubscription.unsubscribe();
        this.moOnDeleteSubscription.unsubscribe();
    };
    DeviceTabProfileComponent.prototype.updateProfileItems = function (device, profile) {
        this.firmwareItems = this.deviceProfileService.getFirmwareItems(device, profile);
        this.softwareItems = this.deviceProfileService.getSoftwareItems(device, profile);
        this.configurationItems = this.deviceProfileService.getConfigurationItems(device, profile);
    };
    DeviceTabProfileComponent.prototype.subscribeToManagedObjects = function () {
        var _this = this;
        this.moOnUpdateSubscription = this.deviceRealtime
            .onUpdate$(this.device.id)
            .subscribe(function (managedObject) {
            _this.updateProfileItems(managedObject, _this.selectedProfile);
        });
        this.moOnDeleteSubscription = this.deviceRealtime.onDelete$(this.device.id).subscribe(function () {
            _this.alertService.danger(gettext('This device has just been deleted. You will be redirected to "All devices" page now.'));
            window.location.href = '#/device';
        });
    };
    DeviceTabProfileComponent.prototype.subscribeToOperations = function () {
        var _this = this;
        var operationsChannel = "/operations/" + this.device.id;
        this.operationsSubscription = this.realtime
            .observable(operationsChannel)
            .pipe(filter(function (_a) {
            var data = _a.data;
            return data.c8y_DeviceProfile;
        }))
            .subscribe(function (_a) {
            var data = _a.data;
            _this.operation = data;
        });
    };
    DeviceTabProfileComponent.ctorParameters = function () { return [
        { type: ManagedObjectRealtimeService },
        { type: DeviceProfileService },
        { type: ActivatedRoute },
        { type: Realtime },
        { type: AlertService }
    ]; };
    DeviceTabProfileComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-tab-profile',
            template: "<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"getDeviceProfilesAndUpdateProfileItems()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'fa-spin': reloading }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n<c8y-action-bar-item [placement]=\"'right'\">\n  <c8y-realtime-btn [service]=\"deviceRealtime\"></c8y-realtime-btn>\n</c8y-action-bar-item>\n\n<div class=\"card card--grid--fullpage card--grid--fullpage card--grid grid__row--2-10--md\">\n  <div class=\"card--grid grid__col--6-6--md\">\n    <!-- AVAILABLE PROFILES -->\n    <div class=\"bg-white\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Device profile</h4>\n      </div>\n      <div class=\"p-16\">\n        <form #deviceProfileForm=\"ngForm\">\n          <div class=\"input-group\">\n            <c8y-typeahead\n              class=\"flex-grow\"\n              name=\"selectProfile\"\n              [(ngModel)]=\"selectedProfile\"\n              placeholder=\"{{ 'Select device profile' | translate }}\"\n              (onSearch)=\"setPipe($event)\"\n              [allowFreeEntries]=\"false\"\n            >\n              <c8y-li\n                *c8yFor=\"let profile of deviceProfiles; pipe: filterPipe\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectProfile(profile); setPipe('')\"\n              >\n                <c8y-highlight\n                  [text]=\"profile.name || '&#45;&#45;'\"\n                  [pattern]=\"pattern\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <div class=\"input-group-btn\">\n              <button\n                class=\"btn btn-primary\"\n                (click)=\"createOperation()\"\n                title=\"{{ 'Assign device profile' | translate }}\"\n                [disabled]=\"!selectedProfile?.id\"\n                translate\n              >\n                Assign device profile\n              </button>\n            </div>\n          </div>\n        </form>\n      </div>\n    </div>\n\n    <!-- INSTALL PROFILE OPERATION -->\n    <div class=\"bg-gray-white\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Currently installed</h4>\n      </div>\n      <div class=\"card-block\">\n        <c8y-single-operation [operation]=\"operation\"></c8y-single-operation>\n      </div>\n    </div>\n  </div>\n  <div class=\"card--grid__inner-scroll flex-col no-align-items\">\n    <div class=\"d-contents\">\n      <!-- FIRMWARE -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Firmware' | translate\"\n        [sectionIcon]=\"'c8y-firmware'\"\n        [emptyStateText]=\"'No firmware to display.' | translate\"\n        [emptyStateDetails]=\"'No firmware assigned.' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"firmwareItems\"\n        [isEmpty]=\"!selectedProfile?.c8y_DeviceProfile?.firmware?.name\"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <div class=\"d-contents\">\n      <!-- SOFTWARE -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Software' | translate\"\n        [sectionIcon]=\"'c8y-tools'\"\n        [emptyStateText]=\"'No software to display.' | translate\"\n        [emptyStateDetails]=\"'No software assigned.' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"softwareItems\"\n        [isEmpty]=\"\n          !selectedProfile?.c8y_DeviceProfile?.software ||\n          selectedProfile?.c8y_DeviceProfile?.software?.length === 0\n        \"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <div class=\"d-contents\">\n      <!-- CONFIGURATION -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Configuration' | translate\"\n        [sectionIcon]=\"'gears'\"\n        [emptyStateText]=\"'No configuration to display' | translate\"\n        [emptyStateDetails]=\"'No configuration assigned' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"configurationItems\"\n        [isEmpty]=\"\n          !selectedProfile?.c8y_DeviceProfile?.configuration ||\n          selectedProfile?.c8y_DeviceProfile?.configuration?.length === 0\n        \"\n        [showTextLabel]=\"false\"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <!-- fill in the remanining vertical space when empty -->\n    <div class=\"card--grid grid__col--6-6--md flex-grow\">\n      <div class=\"bg-white\"></div>\n      <div class=\"bg-gray-white\"></div>\n    </div>\n  </div>\n</div>\n",
            providers: [ManagedObjectRealtimeService]
        })
    ], DeviceTabProfileComponent);
    return DeviceTabProfileComponent;
}());
export { DeviceTabProfileComponent };
